src/go/rdctl/cmd/shutdown.go

```diff
@@ -31,15 +48,144 @@ var shutdownCmd = &cobra.Command{
+				return nil, nil
+			} else {
+				return nil, err
+			}
+		}
+		urlError := new(url.Error)
+		if errors.As(err, &urlError) {
+			return []byte("Rancher Desktop is currently not running (or can't be shutdown via this command)."), nil
+		}
+		return nil, err
+	}
+	if shutdownSettings.DontWait {
+		return output, nil
+	}
+	err = continueShutdown()
+	return output, err
+}
+
+func continueShutdown() error {
+	switch runtime.GOOS {
+	case "darwin":
+		doCheckWithTimeout(checkProcessQemu, pkillQemu, 15, 2, "qemu")
+		doCheckWithTimeout(checkProcessDarwin, pkillDarwin, 5, 1, "the app")
+	case "linux":
+		doCheckWithTimeout(checkProcessQemu, pkillQemu, 15, 2, "qemu")
+		doCheckWithTimeout(checkProcessLinux, pkillLinux, 5, 1, "the app")
+	case "windows":
+		doCheckWithTimeout(checkProcessWindows, factoryreset.PowershellKillWindows, 15, 2, "the app")
+	default:
+		return fmt.Errorf("unhandled runtime: %s", runtime.GOOS)
+	}
+	return nil
+}
+
+func doCheckWithTimeout(checkFunc func() bool, killFunc func(), retryCount int, retryWait int, operation string) {
+	for iter := 0; iter < retryCount; iter++ {
+		if iter > 0 {
+			if shutdownSettings.Verbose {
+				fmt.Fprintf(os.Stderr, "checking %s showed it's still running; sleeping %d seconds\n", operation, retryWait)
+			}
+			time.Sleep(time.Duration(retryWait) * time.Second)
+		}
+		if !checkFunc() {
+			if shutdownSettings.Verbose {
+				fmt.Fprintf(os.Stderr, "%s is no longer running\n", operation)
+			}
+			return
+		}
+	}
+	if shutdownSettings.Verbose {
+		fmt.Fprintf(os.Stderr, "About to force-kill %s\n", operation)
+	}
+	killFunc()
+}
+
+/**
+ * checkProcessX function returns true if it detects the app is still running, false otherwise
+ */
+
+func checkProcessDarwin() bool {
+	return checkProcessLinuxLike("Contents/MacOS/Rancher Desktop")
+}
+
+func checkProcessLinux() bool {
+	return checkProcessLinuxLike("rancher-desktop")
+}
+
+func checkProcessWindows() bool {
+	path, err := factoryreset.GetLockfilePath("rancher-desktop")
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "Error trying to get the lockfile path: %s\n", err)
+		return false
+	}
+	if shutdownSettings.Verbose {
+		fmt.Fprintf(os.Stderr, "GetLockfilePath => %s\n", path)
+	}
+	if _, err = os.Stat(path); err != nil {
+		if !errors.Is(err, os.ErrNotExist) {
+			fmt.Fprintf(os.Stderr, "Error trying to stat %s: %s\n", path, err)
+		}
+		// File either no longer exists or isn't "stat-table"
+		return false
+	}
+	return true
+}
+
+func checkProcessLinuxLike(commandPattern string) bool {
+	result, err := exec.Command("pgrep", "-f", commandPattern).CombinedOutput()
+	if err != nil {
+		return false
+	}
+	ptn, err := regexp.Compile(`\A[0-9\s]+\z`)
+	if err != nil {
+		logrus.WithField("error", err).Warn("failed to compile pattern")
+		return false
+	}
+	return ptn.Match(result)
+}
+
+func checkProcessQemu() bool {
+	return checkProcessLinuxLike("qemu-system")
+}
+
+func pkillQemu() {
+	exec.Command("pkill", "qemu-system").Run()
```

Comment:
There may be qemu from the user outside of Rancher Desktop (e.g. if they're running VMs via virt-manager or something).

Reply:
Yup, I can expand the command-line I match against.